lib/exchanges/gdax.js

```diff
@@ -4,57 +4,71 @@ var request = require('request');
     for (var i = 0; i < count; i++){
-      highbids.set(data.bids[i][2], {
-        rate: parseFloat(data.bids[i][0]),
-        amount: parseFloat(data.bids[i][1])
-      });
-      lowasks.set(data.asks[i][2], {
-        rate: parseFloat(data.asks[i][0]),
-        amount: parseFloat(data.asks[i][1])
-      });
+      orderbook[parseFloat(data.bids[i][0])] = parseFloat(data.bids[i][1]);
+      orderbook[parseFloat(data.asks[i][0])] = -1 * parseFloat(data.asks[i][1]);
     }
 }
 
 // Parse data from Web Socket
 // TODO: Explain (or link to) expected data scheme
 function parse(data){
-  var data = JSON.parse(data)
-
-  // Determine relevant order map to use
-  var bidsOrAsks = data.side === "buy" ? highbids : lowasks;
 
+  var data = JSON.parse(data)
+  var mult = 1;
+  // Simplifying for one orderbook template
+  if (data.side == "sell"){
+    mult = -1;
+  }
+  var key = parseFloat(data.price);
   switch (data.type) {
     case "open":
-      bidsOrAsks.set(data.order_id, {
-        rate: parseFloat(data.price),
-        amount: parseFloat(data.remaining_size)
-      });
+      if(orderbook.hasOwnProperty(key)){
+        orderbook[key] += mult * parseFloat(data.remaining_size);
+      }
+      else {
+        orderbook[key] = mult * parseFloat(data.remaining_size);
+      }
       break;
     case "done":
-      bidsOrAsks.delete(data.order_id);
-      break;
-    case "change":
-      if (bidsOrAsks.get(data.order_id)!= undefined){
-        bidsOrAsks(data.order_id).rate = parseFloat(data.price);
-        bidsOrAsks(data.order_id).amount = parseFloat(data.new_size);
-        bidsOrAsks(data.order_id).amount = parseFloat(data.new_funds);
+      if (orderbook.hasOwnProperty(key)){
+        var tmp = parseFloat(orderbook[key]);
+        delete(orderbook[key]);
+        orderbook[key] = tmp - (mult * parseFloat(data.remaining_size));
+        if ((mult * orderbook[key]) < min){
+          delete(orderbook[key]);
+        }
       }
       break;
-    case "heartbeat":
-    case "received":
     case "match":
-      // TODO: Why do these messages get sent? If we can't do anything with
-      // them remove this TODO and comment the reason why.
+      if (orderbook.hasOwnProperty(key)){
+        var tmp = parseFloat(orderbook[key]);
+
+        delete(orderbook[key]);
+        orderbook[key] = tmp - (mult * parseFloat(data["size"]));
+
+        if ((mult * orderbook[key]) < min){
+          delete(orderbook[key]);
+        }
+      }
       break;
+    case "change":
+      if (orderbook.hasOwnProperty(key)){
+        var tmp = parseFloat(orderbook[key]);
+        delete(orderbook[key]);
+
+        orderbook[key] = tmp - (mult * (parseFloat(data.old_size) - parseFloat(data.new_size)));
+      }
+      break;
+    // These message are for level 4 orderbooks, we only need level 2
     default:
-      console.error("Unexpected data.type!", data.type);
+      //console.error("Unexpected data.type!", data.type);
```

Comment:
It would fire on hb messages and other such data types we ignore. I wasn't sure if just having 3-4 empty switch statements would be the better option.